src/Common/src/System/Text/StringOrCharArray.cs

```diff
@@ -90,6 +92,7 @@ public unsafe bool Equals(StringOrCharArray other)
             }
 
             // CharArray vs CharArray
+            Debug.Assert(CharArray != null && other.CharArray != null);
```

Comment:
For me we don't need to check `other.CharArray != null` which is checked below.
Also seems one of the `String` or `CharArray` initialized with constructor, not both isn't it? in that case this assert could broke test if there is any. I mean CharArray seems not init-ed all the time